rclc/test/rclc/test_executor_multi_threading.cpp

```diff
@@ -0,0 +1,349 @@
+// Copyright (c) 2020 - for information on the respective copyright owner
+// see the NOTICE file and/or the repository https://github.com/ros2/rclc.
+// Copyright 2014 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include <std_msgs/msg/int32.h>
```

Comment:
Double-check include order. First comes header under test, then C standard headers, then C++ standard headers, then probably gtest and then ROS headers.

Reply:
This order did not pass the linter: 
```
#include "rclc/multi_threaded_executor.h"

#include <chrono>
#include <thread>
#include <vector>

#include <gtest/gtest.h>

#include "rcutils/time.h"
#include <std_msgs/msg/int32.h>
#include <example_interfaces/srv/add_two_ints.h>
#include "osrf_testing_tools_cpp/scope_exit.hpp"
```
Error message:
```
3: /home/jst3si/ros_ws/src/rclc/rclc/test/rclc/test_executor_multi_threading.cpp:32:  Found C++ system header after other header. Should be: test_executor_multi_threading.h, c system, c++ system, other.  [build/include_order] [4]
3: /home/jst3si/ros_ws/src/rclc/rclc/test/rclc/test_executor_multi_threading.cpp:33:  Found C++ system header after other header. Should be: test_executor_multi_threading.h, c system, c++ system, other.  [build/include_order] [4]
3: /home/jst3si/ros_ws/src/rclc/rclc/test/rclc/test_executor_multi_threading.cpp:34:  Found C++ system header after other header. Should be: test_executor_multi_threading.h, c system, c++ system, other.  [build/include_order] [4]
3: /home/jst3si/ros_ws/src/rclc/rclc/test/rclc/test_executor_multi_threading.cpp:36:  Found C system header after other header. Should be: test_executor_multi_threading.h, c system, c++ system, other.  [build/include_order] [4]
3: /home/jst3si/ros_ws/src/rclc/rclc/test/rclc/test_executor_multi_threading.cpp:39:  Found C system header after other header. Should be: test_executor_multi_threading.h, c system, c++ system, other.  [build/include_order] [4]
3: /home/jst3si/ros_ws/src/rclc/rclc/test/rclc/test_executor_multi_threading.cpp:40:  Found C system header after other header. Should be: test_executor_multi_threading.h, c system, c++ system, other.  [build/include_order] [4]
3: Category 'build/include_order' errors found: 6
```

I guess the problem is that the "header under test" is in C and not C++. Therefore the linter complains when the standard C++ header comes after the header under test.

Reply:
also this did not work: 
```
#include <gtest/gtest.h>

#include <chrono>
#include <thread>
#include <vector>

#include "rclc/multi_threaded_executor.h"
#include "osrf_testing_tools_cpp/scope_exit.hpp"
#include "rcutils/time.h"
#include <std_msgs/msg/int32.h>
#include <example_interfaces/srv/add_two_ints.h>
```

Error message:
```
3: /home/jst3si/ros_ws/src/rclc/rclc/test/rclc/test_executor_multi_threading.cpp:38:  Found C system header after other header. Should be: test_executor_multi_threading.h, c system, c++ system, other.  [build/include_order] [4]
3: /home/jst3si/ros_ws/src/rclc/rclc/test/rclc/test_executor_multi_threading.cpp:39:  Found C system header after other header. Should be: test_executor_multi_threading.h, c system, c++ system, other.  [build/include_order] [4]
3: Category 'build/include_order' errors found: 2
3: Total errors found: 2
```

Reply:
also this did not work: 
```
#include "rclc/multi_threaded_executor.h"

#include <std_msgs/msg/int32.h>
#include <example_interfaces/srv/add_two_ints.h>
#include <gtest/gtest.h>

#include <chrono>
#include <thread>
#include <vector>

#include "osrf_testing_tools_cpp/scope_exit.hpp"
#include "rcutils/time.h"
```

Error message of linter:
```
3: /home/jst3si/ros_ws/src/rclc/rclc/test/rclc/test_executor_multi_threading.cpp:30:  Found C system header after other header. Should be: test_executor_multi_threading.h, c system, c++ system, other.  [build/include_order] [4]
3: /home/jst3si/ros_ws/src/rclc/rclc/test/rclc/test_executor_multi_threading.cpp:31:  Found C system header after other header. Should be: test_executor_multi_threading.h, c system, c++ system, other.  [build/include_order] [4]
3: /home/jst3si/ros_ws/src/rclc/rclc/test/rclc/test_executor_multi_threading.cpp:32:  Found C system header after other header. Should be: test_executor_multi_threading.h, c system, c++ system, other.  [build/include_order] [4]
3: /home/jst3si/ros_ws/src/rclc/rclc/test/rclc/test_executor_multi_threading.cpp:34:  Found C++ system header after other header. Should be: test_executor_multi_threading.h, c system, c++ system, other.  [build/include_order] [4]
3: /home/jst3si/ros_ws/src/rclc/rclc/test/rclc/test_executor_multi_threading.cpp:35:  Found C++ system header after other header. Should be: test_executor_multi_threading.h, c system, c++ system, other.  [build/include_order] [4]
3: /home/jst3si/ros_ws/src/rclc/rclc/test/rclc/test_executor_multi_threading.cpp:36:  Found C++ system header after other header. Should be: test_executor_multi_threading.h, c system, c++ system, other.  [build/include_order] [4]
3: Category 'build/include_order' errors found: 6

```

Reply:
could not find a different include order, that passes the linter.

Reply:
made a comment about the include order in [3a8ce68](https://github.com/ros2/rclc/pull/87/commits/3a8ce68f1903180e42735d06e53de10f5b4626d7)